---

# Task 1: Familiarize Yourself with the Dataset

There are two main subtasks:  
1. Load the four data files that are part of this dataset.  
2. Explore each dataset and define the data types of each column. Describe the contents of each column, including the nature of the data values. Follow a similar approach as demonstrated in the lecture and tutorial. Discuss the shape of the data, value ranges, and possible categories. Convert any data into a more appropriate format if needed.  

Add as many code and markdown cells as necessary below. Typically, you should perform one type of investigation per code cell with output, and then comment on the results in a markdown cell below it.  

After completing the analysis, write a concise summary in the **Task 1: summary report** cell at the end. The final report should succinctly summarize the main characteristics of each data file, how they relate to each other, and any notable features.


In [2]:
# this is your first code cell, add more cells for code and mardown as needed

### Task 1 Summary Report

Write your report here, summarizing the work done in the cells above.

# Task 2: Data Cleaning, Diagnostics, and Corrections

Identify and resolve issues in the dataset. These can be of different kinds, such as:
- Look for duplicates or implausible values (e.g., users with negative followers, invalid age groups). Thoroughly document your exploration of the data. Decide whether these are outliers or the result of a systemic problem. If the latter, resolve the problem here; if they are outliers, we will address those later.  
- Define at least three heuristics (common-sense rules) to flag unusual entries (e.g., someone watching dozens of streams per day). Discuss whether what you found are errors or artifacts.  

I will repeat the instructions here, but this is expected for each task: each code cell should focus on a specific analysis. You should run the code, and in the markdown below each code cell and its result, provide a detailed interpretation of what you have uncovered in that step. After completing the analyses, summarize the key findings in the indicated summary report cell below.


In [ ]:
# start code for Task 2 here

### Task 2 summary report

# Task 3: Descriptive Statistics and Visualizations

- Create summary statistics and visualizations to understand user characteristics  
- Create summary statistics for streams data overall, and summaries by age group, account tire, or peak viewers  
- Visualize the distribution of streams durations, popular categories,   
- Show category-specific information and compare streamers based on categories  

In [ ]:
#start code for Task 3 here

### Task 3 Summary report
summarize your findings here. 

# Task 4: Dealing with Outliers and Missing Values

- Are there any missing values or outliers in the original data? Perform the analysis and report your findings.  
- Are there any outliers or missing values in the derived data? Perform the analysis and report your findings.  
- If there are no outliers or missing values, introduce 3–5% of missing values in a selected column and 1–2% of outliers in another column.  
- Experiment with two strategies for imputing missing values and two strategies for handling outliers. Show how each strategy affects the characteristics of the data.


In [ ]:
#start code for Task 4 here

### Task 4 Summary report

# Task 5: Correlation and Behavioral Patterns

- Compute pairwise correlations between user features and the number of check-ins  
- Visualize the results using a correlation matrix (use a coolwarm color scheme and proper scaling)  
- Try Spearman rank correlation and explain when it is more appropriate than Pearson

In [ ]:
# tast 5 code starts here

### Task 5 Summary report

# Task 6: Compare Behavior Across User Segments

- Segment users into 2–3 groups (e.g., by age bracket, account tire)  
- Compare their viewing behavior over time using boxplots or dot plots  
- Identify any outliers in group behavior and investigate them

In [ ]:
# task 6 code start here

### Task 6 Summary report

# Task 7: Your Own EDA Questions

Formulate and explore two hypotheses of your own:  
- Propose questions based on curiosity (e.g., “Does a user's variety of categories influence morning vs. evening or session duration?”)  
- Use summary statistics and visualizations to answer each question  
- Briefly discuss what additional data would help you answer the questions more effectively


In [ ]:
# task 7 code start here

### Task 7 Summary report

In [ ]:
# Task 8: EDA Presentation (Insight Report)

Summarize your findings in a polished format:  
- Create 3–4 slides or a single-page report highlighting your most important insights  
- Include at least 2 charts and interpret them using clear, plain language  
- Apply Tufte’s principles: minimize chartjunk, maximize data-to-ink ratio, and ensure clear labeling  

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
viewers = pd.read_csv("viewers.csv")
streamers = pd.read_csv("streamers.csv")
streams = pd.read_csv("streams.csv")
sessions = pd.read_csv("sessions.csv")

In [2]:
datasets = {
    "viewers": viewers,
    "streamers": streamers,
    "streams": streams,
    "sessions": sessions
}
for name, df in datasets.items():
    print(name)
    print("Shape:", df.shape)
    print("Columns:", list(df.columns))
    print()

viewers
Shape: (1000, 6)
Columns: ['viewer_id', 'age_group', 'country', 'account_age_days', 'subscription_tier', 'preferred_category']

streamers
Shape: (76, 7)
Columns: ['streamer_id', 'streamer_name', 'category', 'language', 'partner_status', 'total_followers', 'avg_concurrent_viewers']

streams
Shape: (1199, 9)
Columns: ['stream_id', 'streamer_id', 'started_at', 'ended_at', 'stream_duration_hrs', 'category', 'peak_viewers', 'title_has_hype_word', 'was_raid']

sessions
Shape: (22825, 11)
Columns: ['session_id', 'viewer_id', 'streamer_id', 'stream_id', 'started_at', 'ended_at', 'duration_mins', 'chat_messages_sent', 'bits_cheered', 'followed_during', 'subscribed_during']



### Explanation
This section loads the four Twitch datasets into separate Pandas DataFrames. I first inspect the shape and column names of each table so I can understand what each file contains before cleaning or merging the data.

In [6]:
for name, df in datasets.items():
    print(f"\n{name.upper()}")
    df.info()
    print("\nMissing Values:")
    print(df.isnull().sum())


VIEWERS
<class 'pandas.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype
---  ------              --------------  -----
 0   viewer_id           1000 non-null   str  
 1   age_group           1000 non-null   str  
 2   country             920 non-null    str  
 3   account_age_days    1000 non-null   int64
 4   subscription_tier   1000 non-null   str  
 5   preferred_category  1000 non-null   str  
dtypes: int64(1), str(5)
memory usage: 47.0 KB

Missing Values:
viewer_id              0
age_group              0
country               80
account_age_days       0
subscription_tier      0
preferred_category     0
dtype: int64

STREAMERS
<class 'pandas.DataFrame'>
RangeIndex: 76 entries, 0 to 75
Data columns (total 7 columns):
 #   Column                  Non-Null Count  Dtype
---  ------                  --------------  -----
 0   streamer_id             76 non-null     str  
 1   streamer_name           76 non-nul

### Dataset Overview

Twitch dataset consists of four related tables- `viewers` table contains demographic and account information for 1,000 viewers. `streamers` table contains metadata for 76 streamers, including category, language, and follower statistics. `streams` table records 1,199 individual broadcasts, while the `sessions` table contains 22,825 viewing sessions and serves as the main behavioural event log
 tables are connected through `viewer_id`, `streamer_id`, and `stream_id`. The sessions table links viewers to specific streams and streamers, allowing behavioural data to be combined with viewer and streamer metadata for analysis

In [12]:
print("Duplicate viewers:", viewers.duplicated().sum())
print("Duplicate streamers:", streamers.duplicated().sum())
print("Duplicate streams:", streams.duplicated().sum())
print("Duplicate sessions:", sessions.duplicated().sum())

Duplicate viewers: 0
Duplicate streamers: 0
Duplicate streams: 0
Duplicate sessions: 0


In [13]:
print("Duplicate viewer_id:", viewers["viewer_id"].duplicated().sum())
print("Duplicate streamer_id:", streamers["streamer_id"].duplicated().sum())
print("Duplicate stream_id:", streams["stream_id"].duplicated().sum())
print("Duplicate session_id:", sessions["session_id"].duplicated().sum())

Duplicate viewer_id: 0
Duplicate streamer_id: 0
Duplicate stream_id: 0
Duplicate session_id: 0


### Duplicate Check

I checked for duplicate rows and duplicate primary key values across all four datasets- noduplicates were found indicating that each viewer, streamer, stream, and session record is uniquely identified, which suggests that duplicate records are not a significant data quality issue in the dataset.

In [14]:
sessions[~sessions["viewer_id"].isin(viewers["viewer_id"])].shape

(0, 11)

In [15]:
sessions[~sessions["streamer_id"].isin(streamers["streamer_id"])].shape

(60, 11)

In [16]:
sessions[~sessions["stream_id"].isin(streams["stream_id"])].shape

(35, 11)

In [17]:
streams[~streams["streamer_id"].isin(streamers["streamer_id"])].shape

(0, 9)

In [20]:
streamers.describe()

,total_followers,avg_concurrent_viewers
count,7.600000e+01,76.000000
mean,3.300886e+06,31227.986842
std,3.793772e+06,38307.223652
min,-4.696400e+04,634.000000
25%,7.800000e+05,6633.250000
50%,2.000000e+06,14762.500000
75%,3.875000e+06,36840.750000
max,1.870000e+07,156401.000000


In [21]:
streams.describe()

,stream_duration_hrs,peak_viewers
count,1199.000000,1119.000000
mean,4.625104,38375.658624
std,7.995028,50377.085406
min,1.500000,-50.000000
25%,2.500000,7598.500000
50%,3.000000,20224.000000
75%,5.000000,47941.000000
max,96.000000,370165.000000


In [22]:
sessions.describe()

,duration_mins,chat_messages_sent,bits_cheered
count,22825.000000,22825.000000,22825.000000
mean,59.051303,6.442629,398.425323
std,60.383979,46.576309,1571.528658
min,0.000000,0.000000,0.000000
25%,15.000000,0.000000,0.000000
50%,39.000000,1.000000,0.000000
75%,83.000000,5.000000,0.000000
max,443.000000,1987.000000,10000.000000


In [27]:
streamers[streamers["total_followers"] < 0]

,streamer_id,streamer_name,category,language,partner_status,total_followers,avg_concurrent_viewers
1,STR0002,Pokimane,Just Chatting,English,True,-25551,71660
6,STR0007,Devin_Nash,Just Chatting,English,True,-45130,5114
7,STR0008,Ibai,Just Chatting,Spanish,True,-46964,156401
12,STR0013,Fuslie,Just Chatting,English,True,-8902,9540
24,STR0025,GeT_RiGhT,FPS,English,True,-36135,10785


In [28]:
streams[streams["peak_viewers"] < 0]

,stream_id,streamer_id,started_at,ended_at,stream_duration_hrs,category,peak_viewers,title_has_hype_word,was_raid
30,SM00031,STR0002,2024-02-23 18:00:00,2024-02-23 20:30:00,2.5,Just Chatting,-19.0,False,False
58,SM00059,STR0004,2024-03-22 11:00:00,2024-03-22 16:00:00,5.0,Just Chatting,-41.0,True,False
124,SM00125,STR0008,2024-03-10 14:00:00,2024-03-10 18:00:00,4.0,Just Chatting,-43.0,False,False
145,SM00146,STR0009,2024-03-22 15:00:00,2024-03-22 22:00:00,7.0,Just Chatting,-24.0,True,False
174,SM00175,STR0011,2024-01-06 16:00:00,2024-01-06 21:00:00,5.0,Just Chatting,-23.0,False,False
182,SM00183,STR0012,2024-02-04 19:00:00,2024-02-04 23:00:00,4.0,Just Chatting,-36.0,False,False
197,SM00198,STR0013,2024-03-14 10:00:00,2024-03-14 15:00:00,5.0,Just Chatting,-50.0,True,False
227,SM00228,STR0015,2024-01-22 10:00:00,2024-01-22 15:00:00,5.0,Just Chatting,-5.0,True,False
239,SM00240,STR0016,2024-01-04 11:00:00,2024-01-04 18:00:00,7.0,Just Chatting,-3.0,False,False
244,SM00245,STR0016,2024-01-23 12:00:00,2024-01-23 19:00:00,7.0,Just Chatting,-50.0,False,False


In [29]:
streams[streams["stream_duration_hrs"] > 24]

,stream_id,streamer_id,started_at,ended_at,stream_duration_hrs,category,peak_viewers,title_has_hype_word,was_raid
6,SM00007,STR0001,2024-03-05 12:00:00,2024-03-05 16:00:00,48.0,Just Chatting,NaN,True,False
48,SM00049,STR0003,2024-02-02 10:00:00,2024-02-02 15:00:00,72.0,Just Chatting,21516.0,False,False
284,SM00285,STR0018,2024-01-24 14:00:00,2024-01-24 17:00:00,72.0,FPS,69221.0,False,False
332,SM00333,STR0021,2024-02-12 22:00:00,2024-02-13 02:00:00,72.0,FPS,120277.0,False,False
357,SM00358,STR0023,2024-03-16 09:00:00,2024-03-16 12:00:00,48.0,FPS,6073.0,False,True
364,SM00365,STR0024,2024-02-28 22:00:00,2024-02-29 02:00:00,72.0,FPS,37056.0,True,False
372,SM00373,STR0024,2024-01-18 15:00:00,2024-01-18 22:00:00,48.0,FPS,16570.0,True,False
417,SM00418,STR0026,2024-03-15 21:00:00,2024-03-16 02:00:00,72.0,FPS,34674.0,False,False
458,SM00459,STR0029,2024-02-25 21:00:00,2024-02-25 23:30:00,48.0,FPS,42620.0,False,False
546,SM00547,STR0034,2024-02-25 12:00:00,2024-02-25 16:00:00,96.0,MOBA,4750.0,False,True


In [30]:
sessions.sort_values("duration_mins", ascending=False).head(10)

,session_id,viewer_id,streamer_id,stream_id,started_at,ended_at,duration_mins,chat_messages_sent,bits_cheered,followed_during,subscribed_during
2848,SES002849,VWR00895,STR0076,SM01193,2024-02-12 11:28:00,2024-02-12 18:51:00,443,14,0,False,False
3728,SES003729,VWR00921,STR0008,SM00128,2024-02-07 20:32:00,2024-02-08 03:51:00,439,41,0,False,False
9868,SES009869,VWR00082,STR0031,SM00502,2024-01-24 17:13:00,2024-01-25 00:29:00,436,2,0,False,False
14924,SES014925,VWR00899,STR0037,SM00601,2024-03-05 09:23:00,2024-03-05 16:37:00,434,1,0,False,False
5927,SES005928,VWR00813,STR0042,SM00666,2024-03-01 19:33:00,2024-03-02 02:47:00,434,94,0,False,False
112,SES000113,VWR00844,STR0061,SM00953,2024-01-03 08:09:00,2024-01-03 15:12:00,423,42,0,False,False
5482,SES005483,VWR00942,STR0075,SM01184,2024-03-12 08:52:00,2024-03-12 15:54:00,422,14,0,False,False
7188,SES007189,VWR00390,STR0025,SM00382,2024-01-20 13:00:00,2024-01-20 19:57:00,417,61,0,False,False
2298,SES002299,VWR00539,STR0068,SM01072,2024-03-14 09:03:00,2024-03-14 15:57:00,414,145,779,False,False
5804,SES005805,VWR00869,STR0076,SM01193,2024-02-12 11:37:00,2024-02-12 18:29:00,412,18,0,False,False


# Data Detective Findings
## Issue #1
File: streamers.csv
Column: total_followers
Problem:
Negative follower count detected for Pokimane, Follower counts cannot be negative
Detection:
streamers[streamers["total_followers"] < 0]
Status:
Confirmed

In [7]:
viewers["country"].isnull().sum()

np.int64(80)

In [8]:
streamers["language"].isnull().sum()

np.int64(8)

In [9]:
streams["category"].isnull().sum()

np.int64(50)

In [10]:
streams["peak_viewers"].isnull().sum()

np.int64(80)

In [11]:
sessions["streamer_id"].isnull().sum()

np.int64(60)

Issue #6 — Orphaned Streamer References

In [18]:
sessions[~sessions["streamer_id"].isin(streamers["streamer_id"])].shape

(60, 11)

In [19]:
sessions[~sessions["stream_id"].isin(streams["stream_id"])].shape

(35, 11)

In [24]:
streamers[streamers["total_followers"] < 0]

,streamer_id,streamer_name,category,language,partner_status,total_followers,avg_concurrent_viewers
1,STR0002,Pokimane,Just Chatting,English,True,-25551,71660
6,STR0007,Devin_Nash,Just Chatting,English,True,-45130,5114
7,STR0008,Ibai,Just Chatting,Spanish,True,-46964,156401
12,STR0013,Fuslie,Just Chatting,English,True,-8902,9540
24,STR0025,GeT_RiGhT,FPS,English,True,-36135,10785


In [25]:
streams[streams["peak_viewers"] < 0]

,stream_id,streamer_id,started_at,ended_at,stream_duration_hrs,category,peak_viewers,title_has_hype_word,was_raid
30,SM00031,STR0002,2024-02-23 18:00:00,2024-02-23 20:30:00,2.5,Just Chatting,-19.0,False,False
58,SM00059,STR0004,2024-03-22 11:00:00,2024-03-22 16:00:00,5.0,Just Chatting,-41.0,True,False
124,SM00125,STR0008,2024-03-10 14:00:00,2024-03-10 18:00:00,4.0,Just Chatting,-43.0,False,False
145,SM00146,STR0009,2024-03-22 15:00:00,2024-03-22 22:00:00,7.0,Just Chatting,-24.0,True,False
174,SM00175,STR0011,2024-01-06 16:00:00,2024-01-06 21:00:00,5.0,Just Chatting,-23.0,False,False
182,SM00183,STR0012,2024-02-04 19:00:00,2024-02-04 23:00:00,4.0,Just Chatting,-36.0,False,False
197,SM00198,STR0013,2024-03-14 10:00:00,2024-03-14 15:00:00,5.0,Just Chatting,-50.0,True,False
227,SM00228,STR0015,2024-01-22 10:00:00,2024-01-22 15:00:00,5.0,Just Chatting,-5.0,True,False
239,SM00240,STR0016,2024-01-04 11:00:00,2024-01-04 18:00:00,7.0,Just Chatting,-3.0,False,False
244,SM00245,STR0016,2024-01-23 12:00:00,2024-01-23 19:00:00,7.0,Just Chatting,-50.0,False,False


In [26]:
streams[streams["stream_duration_hrs"] > 24]

,stream_id,streamer_id,started_at,ended_at,stream_duration_hrs,category,peak_viewers,title_has_hype_word,was_raid
6,SM00007,STR0001,2024-03-05 12:00:00,2024-03-05 16:00:00,48.0,Just Chatting,NaN,True,False
48,SM00049,STR0003,2024-02-02 10:00:00,2024-02-02 15:00:00,72.0,Just Chatting,21516.0,False,False
284,SM00285,STR0018,2024-01-24 14:00:00,2024-01-24 17:00:00,72.0,FPS,69221.0,False,False
332,SM00333,STR0021,2024-02-12 22:00:00,2024-02-13 02:00:00,72.0,FPS,120277.0,False,False
357,SM00358,STR0023,2024-03-16 09:00:00,2024-03-16 12:00:00,48.0,FPS,6073.0,False,True
364,SM00365,STR0024,2024-02-28 22:00:00,2024-02-29 02:00:00,72.0,FPS,37056.0,True,False
372,SM00373,STR0024,2024-01-18 15:00:00,2024-01-18 22:00:00,48.0,FPS,16570.0,True,False
417,SM00418,STR0026,2024-03-15 21:00:00,2024-03-16 02:00:00,72.0,FPS,34674.0,False,False
458,SM00459,STR0029,2024-02-25 21:00:00,2024-02-25 23:30:00,48.0,FPS,42620.0,False,False
546,SM00547,STR0034,2024-02-25 12:00:00,2024-02-25 16:00:00,96.0,MOBA,4750.0,False,True
